pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +131,477 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	nodeName := node.ObjectMeta.Name
+
+	// handling of lease as a prerequisite to state transitions.
+	leaseStatus, err, lease := r.doesLeaseExist(r.client, nodeName)
+	switch leaseStatus {
+	case LeaseStatusExpired:
+		{
+			if err := r.client.Delete(context.TODO(), lease); err != nil {
+				r.reqLogger.Infof("failed to delete expired lease. node: %s err: %v\n", nodeName, err)
+
+			}
+		}
+		fallthrough
+
+	case LeaseStatusNotFound:
+		{
+			if err := r.createLease(node, specAnnoData.RequestedMaintenanceWindowSeconds); err != nil {
+				statusAnnoData.Status = NodeStateFailedToCreateLease
+				_, err = r.setStatusAnnotation(node, statusAnnoData)
+				return true, &reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+			}
+		}
 
-	r.drainer = &drain.Helper{}
+	case LeaseStatusActiveOwnedByMe:
+		{
 
-	//Continue even if there are pods not managed by a ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet.
-	//This is required because VirtualMachineInstance pods are not owned by a ReplicaSet or DaemonSet controller.
-	//This means that the drain operation can’t guarantee that the pods being terminated on the target node will get
-	//re-scheduled replacements placed else where in the cluster after the pods are evicted.
-	//KubeVirt has its own controllers which manage the underlying VirtualMachineInstance pods.
-	//Each controller behaves differently to a VirtualMachineInstance being evicted.
-	r.drainer.Force = true
+			r.reqLogger.Infof("holding lease. Maintenance mode ongoing. node: %s\n", nodeName)
+		}
 
-	//Continue even if there are pods using emptyDir (local data that will be deleted when the node is drained).
-	//This is necessary for removing any pod that utilizes an emptyDir volume.
-	//The VirtualMachineInstance Pod does use emptryDir volumes,
-	//however the data in those volumes are ephemeral which means it is safe to delete after termination.
-	r.drainer.DeleteLocalData = true
+	case LeaseStatusActiveOwnedByDifferentHolder:
+		{
+			if statusAnnoData.Status == NodeStateInit {
+				r.reqLogger.Infof("wait to acquire lease. requeue. node: %s\n", nodeName)
+				r.setStatusAnnotation(node, statusAnnoData)
+				return true, &reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+			}
 
-	//Ignore DaemonSet-managed pods.
-	//This is required because every node running a VirtualMachineInstance will also be running our helper DaemonSet called virt-handler.
-	//This flag indicates that it is safe to proceed with the eviction and to just ignore DaemonSets.
-	r.drainer.IgnoreAllDaemonSets = true
+			statusAnnoData.Status = NodeStateLeaseNotOwnedByNode
+			r.setStatusAnnotation(node, statusAnnoData)
 
-	//Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.
-	r.drainer.GracePeriodSeconds = -1
+			err := fmt.Errorf("lease got acquired by someone while request was active. node: %s", nodeName)
```

Comment:
This is not an error.
You would end up here if someone already had the lease prior to the maintenance window being requested

Reply:
Oh, I see, you're differentiating between the two cases based on `statusAnnoData.Status`
Probably reasonable, but maybe we should go back to a "waiting for the lease" state.

Reply:
bump